### DISCUSSION TITLE
Paint it Black

### DISCUSSION THREAD

**Discussion by jangop** [sylikc/pyexiftool#23] [General]
Any opinions on standardizing the existing code base? Or, rather, a lack of opinions?

I feel that it would make contributing much easier and to avoid wasting anything on discussions, my suggestion is to include a github workflow to run [Black](https://github.com/psf/black).

**Comment by sylikc** [comment]
Hmm, interesting... thanks for introducing such a transformative project!

I went through a light scour of the documentation this weekend and wasn't a big fan of some of the things it does.  I guess it works out well for bigger projects which have tons of programmers who do things different ways, but I really wished it was more configurable.

I'll look into it further, and test its outputs, but the strict adherence to beautifying the code I think actually makes it harder to read.

Like it has a sample
```python
def very_important_function(
    template: str,
    *variables,
    file: os.PathLike,
    engine: str,
    header: bool = True,
    debug: bool = False,
):
    """Applies `variables` to the `template` and writes to `file`."""
    with open(file, "w") as f:
        ...
```

I prefer having TWO spaces for the function definition to show at a glance that it is on a separate level to the body of the function.

Their whole quote thing is something that seems to make things more unreadable for the sake of being pedantic.  Black standardizes all single quotes into double quotes.  I actually thought that it was a great thing that Python allowed using both, because it makes strings more readable.  If I have double quotes in my string, I can use single quotes, and vice versa, so I can leave the backslashes at home.


### PyExifTool does need some styling

Anyways, that's not to say that the code doesn't need styling cleanup.  The docstrings for one are a mess right now.  My contributions to the project, documentation style and stuff leaves lots to be desired.  

I'd prefer to get the code more functional (and not brittle), and documentation more complete, before improving the styling

**Reply by jangop** [reply]
Alright, so, everyone goes down this rabbit hole at some point or another. 😉 While I personally also prefer typing single quotes, I love no longer having to worry about it.

Of course, you can also go for [pycodestyle](https://pypi.org/project/pycodestyle). Anything that lets me check my code before committing. It was painful [modifying `set_tags` and leaving the camelcased arguments](https://github.com/sylikc/pyexiftool/pull/19/commits/98357b75c0387c3a6dc4adb119bd58db8878bc62).

Lastly, while I applaud a focus on functionality and documentation, let me remark that styling should be done as early as possible to minimize the effects on code's history.

**Reply by sylikc** [reply]
with regards to just the camelcasing in `set_tags` lol that's me merging a whole bunch of random PR from the old repo... I certainly didn't write that code haha.  **For the Helper, I haven't had much time to update it, but probably take it all back to the Python code style of lower_lower_lower**

I do really <3 tabs though ... as a fact that in whatever editor I choose, I can make tabs appear any number of spaces.

Let me look at `pycodestyle`.  A quick documentation skim, I like it more than Black already... I like that it warns me instead of changing the code... that way I can decide which of the PEP8 I want to actually implement.

**Reply by sylikc** [reply]
UPDATE: I'm testing `flake8` with `pep8-naming` plugin (deal with _camelCaseStuff_) and `pycodestyle` to see how this works out.  If it seems to be good, then I'll have a script in the `scripts` folder that will run it with some standard config file.

I'll see how close I can get the code to PEP8.  Not totally a fan of some of the stuff, like lines at 79 characters ...

**Comment by jangop** [comment]
So, ordinarily you would set up a github workflow, like so: https://github.com/jangop/pyexiftool/blob/master/.github/workflows/lint-and-test.yml

Then, everything gets checked automatically and you get awesome notifications such as: https://github.com/jangop/pyexiftool/runs/2702057786
